lib/ex_unit/lib/ex_unit/server.ex

```diff
@@ -120,18 +133,23 @@ defmodule ExUnit.Server do
     state
   end
 
-  defp take_modules(%{waiting: {from, _count}, async_modules: [], loaded: :done} = state) do
-    GenServer.reply(from, nil)
-    %{state | waiting: nil}
-  end
+  defp take_modules(%{waiting: {from, count}} = state) do
+    has_async_modules? = not :queue.is_empty(state.async_modules)
 
-  defp take_modules(%{async_modules: []} = state) do
-    state
-  end
+    cond do
+      not has_async_modules? and state.loaded == :done ->
+        GenServer.reply(from, nil)
+        %{state | waiting: nil}
 
-  defp take_modules(%{waiting: {from, count}, async_modules: modules} = state) do
-    {reply, modules} = Enum.split(modules, count)
-    GenServer.reply(from, reply)
-    %{state | async_modules: modules, waiting: nil}
+      not has_async_modules? ->
+        state
+
+      true ->
+        n = :queue.len(state.async_modules)
+        count = min(count, n)
+        {modules, async_modules} = :queue.split(count, state.async_modules)
```

Comment:
Length is an expensive operation (linear), what we probably want to do is a "take_until" or "take_while". I have implemented this 10 times in GenStage projects, let me try to find one.

Reply:
I couldn't find it, sorry. But the idea is to have a loop that pops from the queue, either until nothing is popped or we reach 0. :)

Reply:
Understood 👍